Tactics/Sym.lean

```diff
@@ -409,3 +410,39 @@ Did you remember to generate step theorems with:
     -- The main loop
     for _ in List.range n do
       c ← sym1 c whileTac
+
+    -- Check if we can substitute the final state
+    if c.runSteps? = some 0 then
+      let msg := do
+        let hRun ← userNameToMessageData c.h_run
+        pure m!"runSteps := 0, substituting along {hRun}"
+      withTraceNode `Tactic.sym (fun _ => msg) <| withMainContext do
+        let s ← SymContext.findFromUserName c.state
+        let sfEq ← mkEq s.toExpr c.finalState
+
+        let goal ← getMainGoal
+        trace[Tactic.sym] "original goal:\n{goal}"
+        let ⟨hEqId, goal⟩ ← do
+          let hRun ← SymContext.findFromUserName c.h_run
+          goal.note `this (← mkEqSymm hRun.toExpr) sfEq
+        goal.withContext <| do
+          trace[Tactic.sym] "added {← userNameToMessageData `this} of type \
+            {sfEq} in:\n{goal}"
+
+        let goal ← subst goal hEqId
+        trace[Tactic.sym] "performed subsitutition in:\n{goal}"
+
+        replaceMainGoal [goal]
+
+    -- Rudimentary aggregation: we feed all the axiomatic effect hypotheses
+    -- added while symbolically evaluating to `simp`
+    withMainContext <| do
```

Comment:
I see that aggregation indeed happens only when we reach the final state because the theorem's conclusion is typically in its terms (anyway the user doesn't have a way to refer to the intermediate states when stating the theorem). `axHyps` won't help in simplifying the conclusion unless the effects of the last instruction are also in the context.

Reply:
We could do some rudimentary intermediate aggregation, by doing a `simp_all` instead of only simplifying the goal. That feels like it could be much more brittle, though, which is why I shied away from that approach